### PR TITLE
feat: add sys.stack_trace system table to dump druid nodes threads

### PR DESCRIPTION
**Pull request by FrankChen021**
## Summary

- add the `sys.stack_trace` system table for collecting live Java platform-thread snapshots from explicitly selected Druid nodes
- add the `/status/stack` endpoint and a stateless collector with jstack-style formatting, CPU timing, lock, and deadlock information
- support the `maxStackTraceFrameDepth` SQL query context and HTTP parameter with a default of 100 and a valid range of 10–1000
- document the table, endpoint, conversion behavior, and virtual-thread limitation

## Motivation

Expose targeted runtime diagnostics through SQL so operators and diagnostic AI agents can investigate thread contention, deadlocks, and CPU-heavy threads without collecting stack traces from an entire cluster unintentionally.

## Validation

image

image

- `StatusResourceTest` (10 tests)
- `SystemSchemaTest` (28 tests)
- `SystemStackTraceTableTest` embedded HTTP/SQL coverage (7 tests)
- `CalciteQueryTest#testInformationSchemaTables`
- Checkstyle for `server`, `sql`, and `embedded-tests`
- documentation spellcheck
- `git diff --check`
- Tested on a test cluter

### CODE REVIEW

File: server/src/main/java/org/apache/druid/server/StackTraceCollector.java [hunk 1]
```diff
@@ -0,0 +1,424 @@
...
+    public Long getCpuTimeNs()
+    {
+      return cpuTimeNs;
+    }
+
+    @JsonProperty
+    public Long getUserCpuTimeNs()
+    {
+      return userCpuTimeNs;
+    }
```
[the remark this answers is not in the document]
Comment (FrankChen021) [reply server/src/main/java/org/apache/druid/server/StackTraceCollector.java:394]: Addressed in commit d7b6588cb5. Added @Nullable to getCpuTimeNs, getUserCpuTimeNs, getLockName, getLockOwnerId, and getLockOwnerName so the getter contracts match their nullable backing fields.

File: sql/src/main/java/org/apache/druid/sql/calcite/schema/SystemStackTraceTable.java [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+        StringUtils.format(
+            "/status/stack?%s=%d",
+            StackTraceCollector.MAX_STACK_TRACE_FRAME_DEPTH_KEY,
+            maxStackTraceFrameDepth
+        )
+    ).toString();
+    try {
+      final Request request = new Request(HttpMethod.GET, URI.create(url).toURL());
+      final StringFullResponseHolder response = httpClient
+          .go(request, new StringFullResponseHandler(StandardCharsets.UTF_8))
```
Comment (FrankChen021) [inline sql/src/main/java/org/apache/druid/sql/calcite/schema/SystemStackTraceTable.java]: [P2] Cancel the HTTP request when SQL execution is interrupted

Calling `.get()` directly discards the request future. If a `sys.stack_trace` query is canceled while a node is returning a large or deep snapshot, the catch block interrupts the caller but leaves the underlying HTTP request running and buffering until completion or timeout. Retain the future and use `FutureUtils.get(future, true)`, or explicitly cancel it in the interruption path.

**Review by FrankChen021 — COMMENTED** [review]
| Severity | Findings |
| --- | ---: |
| P0 | 0 |
| P1 | 0 |
| P2 | 1 |
| P3 | 0 |
| Total | 1 |

| Severity | Findings |
| --- | ---: |
| P0 | 0 |
| P1 | 0 |
| P2 | 1 |
| P3 | 0 |
| Total | 1 |

Reviewed 12 of 12 changed files. Found one cancellation and resource-lifecycle issue in the remote stack-trace request path.

---

This is an automated review by Codex GPT-5.6-Sol

File: sql/src/main/java/org/apache/druid/sql/calcite/schema/SystemStackTraceTable.java (hunk 2 above)
Comment (FrankChen021) [reply sql/src/main/java/org/apache/druid/sql/calcite/schema/SystemStackTraceTable.java]: Addressed in commit c6365d775e. The HTTP client future is now passed to FutureUtils.get(..., true), which cancels it when SQL execution is interrupted. Added testStackTraceTable_withInterruptedException to verify cancellation and interrupt preservation.

**Comment by jtuglu1** [comment]
Can you outline the benefits of this approach compared to simply running perf, etc on the Druid process itself? Is this supposed to mirror: https://clickhouse.com/docs/reference/system-tables/trace_log?

**Comment by FrankChen021** [comment]
> Can you outline the benefits of this approach compared to simply running perf, etc on the Druid process itself? Is this supposed to mirror: https://clickhouse.com/docs/reference/system-tables/trace_log?

It's not similar to the trace_log in clickhouse, but similar to the system.stack_trace table in clickhouse.
This provides an easier and simpler way to dump a stack trace to analyze thread usage/potential contention.
With this SQL support, AI agents can get thread dump to do the analysis for us.

**Review by FrankChen021 — COMMENTED** [review]
| Severity | Findings |
| --- | ---: |
| P0 | 0 |
| P1 | 0 |
| P2 | 1 |
| P3 | 0 |
| Total | 1 |

Reviewed 12 of 12 changed files. Found one P2 issue: the updated code marks the future cancelled, but does not abort the production Netty HTTP exchange, so interrupted queries can retain transport resources.

---

This is an automated review by Codex GPT-5.6-Sol

File: sql/src/main/java/org/apache/druid/sql/calcite/schema/SystemStackTraceTable.java [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+    final String url = druidNode.getUriToUse().resolve(
+        StringUtils.format(
+            "/status/stack?%s=%d",
+            StackTraceCollector.MAX_STACK_TRACE_FRAME_DEPTH_KEY,
+            maxStackTraceFrameDepth
+        )
+    ).toString();
+    try {
+      final Request request = new Request(HttpMethod.GET, URI.create(url).toURL());
+      final StringFullResponseHolder response = FutureUtils.get(
```
Comment (FrankChen021) [inline sql/src/main/java/org/apache/druid/sql/calcite/schema/SystemStackTraceTable.java:238]: [P2] Abort the HTTP transport, not just its future

FutureUtils.get(..., true) only calls cancel(true) on the returned future. The production NettyHttpClient.go path returns a plain SettableFuture without wiring cancellation to close the channel or return its pooled resource, while StringFullResponseHandler continues buffering chunks. Interrupting the SQL query therefore still leaves the remote request and pooled connection active until the node responds or the default read timeout expires. The new test verifies only the SettableFuture's cancelled flag, not transport teardown. Please make interruption abort the underlying exchange.

Comment (FrankChen021) [reply sql/src/main/java/org/apache/druid/sql/calcite/schema/SystemStackTraceTable.java:238]: Addressed in commit b1142d4c03. NettyHttpClient now observes cancellation of the returned request future, stops processing further response messages, closes the underlying channel, and returns the pooled resource after the close completes. Added FriendlyServersTest.testCancelRequestClosesConnection, which verifies that the server observes the socket close and that a subsequent request succeeds using the same client (covering pooled-resource return). Verified the focused transport test, the full FriendlyServersTest class, and SystemSchemaTest#testStackTraceTable_withInterruptedException on JDK 25.

**Review by FrankChen021 — COMMENTED** [review]
I have reviewed the updated cancellation path for correctness, transport cleanup, pooled-resource lifecycle, and integration risks; no issues found. The prior Netty transport-cancellation finding is resolved.

Reviewed 14 of 14 changed files.

---

This is an automated review by Codex GPT-5.6-Sol

**Review by FrankChen021 — COMMENTED** [review]
I have reviewed the code for correctness, edge cases, concurrency, and integration risks; no issues found.

Reviewed 15 of 15 changed files.

Validation: `git diff --check` passed. Builds and tests were not run.



---

This is an automated review by Codex GPT-5.6-Luna(max)

**Comment by FrankChen021** [comment]
This needs to be rebased on top of #20183 so that it's served by native query engine

**Review by FrankChen021 — COMMENTED** [review]
I have reviewed the code for correctness, edge cases, concurrency, and integration risks; no issues found.

Reviewed 15 of 15 changed files.



---

This is an automated review by Codex GPT-5.6-Luna(max)
